src/Razor/Razor.Runtime/src/Runtime/TagHelpers/TagHelperRunner.cs

```diff
@@ -20,42 +20,65 @@ public class TagHelperRunner
         /// </param>
         /// <returns>Resulting <see cref="TagHelperOutput"/> from processing all of the
         /// <paramref name="executionContext"/>'s <see cref="ITagHelper"/>s.</returns>
-        public async Task RunAsync(TagHelperExecutionContext executionContext)
+        public Task RunAsync(TagHelperExecutionContext executionContext)
         {
             if (executionContext == null)
             {
                 throw new ArgumentNullException(nameof(executionContext));
             }
 
             var tagHelperContext = executionContext.Context;
+            var tagHelpers = executionContext.TagHelpers;
+            OrderTagHelpers(tagHelpers);
 
-            OrderTagHelpers(executionContext.TagHelpers);
-
-            for (var i = 0; i < executionContext.TagHelpers.Count; i++)
+            // Read interface .Count once rather than per iteration
+            var count = tagHelpers.Count;
+            for (var i = 0; i < count; i++)
             {
-                executionContext.TagHelpers[i].Init(tagHelperContext);
+                tagHelpers[i].Init(tagHelperContext);
             }
 
             var tagHelperOutput = executionContext.Output;
 
-            for (var i = 0; i < executionContext.TagHelpers.Count; i++)
+            for (var i = 0; i < count; i++)
             {
-                await executionContext.TagHelpers[i].ProcessAsync(tagHelperContext, tagHelperOutput);
+                var task = tagHelpers[i].ProcessAsync(tagHelperContext, tagHelperOutput);
+                if (!task.IsCompletedSuccessfully)
+                {
+                    return Awaited(task, executionContext, i, count);
```

Comment:
Is this right? don't we need to increment `i`. it looks like this will initialize the `i`th entry twice if there are > `i` entries.

Reply:
Good catch